# Part 3: Machine Learning

### Import Libaries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data=pd.read_csv('../Data/smartwatch_data_ready_for_ml.csv')

## 3.1: Training Machine Learning Model

### 3.1.1: Split data to X_train, X_valid, y_train, y_valid

In [3]:
from sklearn.model_selection import train_test_split

X=data.drop(columns=['price', 'price_log'])
y=data['price_log']

X_train, X_valid, y_train, y_valid = train_test_split(X, y, train_size=0.8, test_size=0.2, random_state=0)

In [4]:
X_train.shape

(2831, 20)

In [5]:
X_valid.shape

(708, 20)

In [6]:
X.shape

(3539, 20)

### 3.1.2: Linear Regression Model

In [7]:
from sklearn.linear_model import LinearRegression

lr_model=LinearRegression()
lr_model.fit(X_train, y_train)
y_preds=lr_model.predict(X_valid)


In [8]:
y_preds

array([6.33866419, 6.2552285 , 6.09024872, 6.08806736, 6.2610322 ,
       6.27663692, 6.23869817, 6.15470255, 6.20388635, 6.28840907,
       6.32256829, 6.41704531, 6.41716209, 6.25490605, 6.27615676,
       6.18137775, 6.58986401, 6.19728004, 6.39183902, 6.54418647,
       6.625096  , 6.27973311, 6.32652386, 6.34920726, 6.26569622,
       6.25043235, 6.62114044, 6.27832735, 6.39027602, 6.15477861,
       6.21698298, 6.32652386, 6.26569622, 6.40151804, 6.48547302,
       6.32906263, 6.14669497, 6.28840907, 6.4192129 , 6.28228292,
       6.33091461, 6.23507082, 6.22501111, 6.33215922, 6.13543158,
       6.48405249, 6.33091461, 6.30503595, 6.25707663, 6.38664661,
       6.57484114, 6.41152674, 6.47792634, 6.2832447 , 6.4192129 ,
       6.25510096, 6.46096842, 6.13779795, 6.39342204, 6.32256829,
       6.27756253, 6.38475745, 6.32100529, 6.43234695, 6.30765461,
       6.51162018, 6.19891009, 6.38762971, 6.26366385, 6.6128437 ,
       6.32652386, 6.10286529, 6.03695296, 6.24977906, 6.29689

In [9]:
pd.DataFrame ({'y': y_valid.head(), 'y_pred': y_preds[:5]})

,y,y_pred
2606,6.198804,6.338664
2010,6.311735,6.255229
2804,5.802088,6.090249
1632,6.346075,6.088067
2372,6.216606,6.261032


In [10]:
y_valid_usd = np.expm1(y_valid)
y_pred_usd = np.expm1(y_preds)
lr_comparison_df = pd.DataFrame({
    'Actual Price ($)': y_valid_usd,
    'Predicted Price ($)': y_pred_usd,
    'Difference ($)': y_valid_usd - y_pred_usd
})

print(lr_comparison_df.head())

      Actual Price ($)  Predicted Price ($)  Difference ($)
2606            491.16           565.039686      -73.879686
2010            550.00           519.728348       30.271652
2804            329.99           440.531215     -110.541215
1632            569.25           439.569128      129.680872
2372            500.00           522.759284      -22.759284


### 3.1.3: Random Forest Model

In [11]:
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor

rf_model=RandomForestRegressor(random_state=0)
rf_model.fit(X_train, y_train)

RandomForestRegressor(random_state=0)

In [12]:
rf_preds=rf_model.predict(X_valid)

In [13]:
rf_preds[:5]

array([6.19629576, 6.21326421, 5.81507389, 6.27095487, 6.20199655])

In [14]:
pd.DataFrame ({'y': y_valid.head(), 'y_pred': rf_preds[:5]})

,y,y_pred
2606,6.198804,6.196296
2010,6.311735,6.213264
2804,5.802088,5.815074
1632,6.346075,6.270955
2372,6.216606,6.201997


In [15]:
rf_preds_usd = np.expm1(rf_preds)
rf_comparison_df = pd.DataFrame({
    'Actual Price ($)': y_valid_usd,
    'Predicted Price ($)': rf_preds_usd,
    'Difference ($)': y_valid_usd - rf_preds_usd
})

print(rf_comparison_df.head())

      Actual Price ($)  Predicted Price ($)  Difference ($)
2606            491.16           489.927157        1.232843
2010            550.00           498.328505       51.671495
2804            329.99           334.316174       -4.326174
1632            569.25           527.982247       41.267753
2372            500.00           492.733822        7.266178


### 3.1.4: XGBoost Model

In [16]:
import xgboost as xgb
xgb_model = xgb.XGBRegressor(random_state=0)
xgb_model.fit(X_train, y_train)
xgb_preds = xgb_model.predict(X_valid)

xgb_preds[:5]
pd.DataFrame({'y': y_valid.head(), 'y_pred': xgb_preds[:5]})

xgb_preds_usd = np.expm1(xgb_preds)
xgb_comparison_df = pd.DataFrame({
    'Actual Price ($)': y_valid_usd,
    'Predicted Price ($)': xgb_preds_usd,
    'Difference ($)': y_valid_usd - xgb_preds_usd
})
print(xgb_comparison_df.head())

      Actual Price ($)  Predicted Price ($)  Difference ($)
2606            491.16           484.655121        6.504879
2010            550.00           504.046204       45.953796
2804            329.99           330.152008       -0.162008
1632            569.25           528.063843       41.186157
2372            500.00           508.847321       -8.847321


## 3.2: Model Evaluation

In [17]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

def get_metrics(y_true_log, y_pred_log):
    y_true = np.expm1(y_true_log)
    y_pred = np.expm1(y_pred_log)
    return {
        'MAE': mean_absolute_error(y_true, y_pred),
        'RMSE': np.sqrt(mean_squared_error(y_true, y_pred)),
        'R2': r2_score(y_true, y_pred)
    }
results = pd.DataFrame({
    'Linear Regression': get_metrics(y_valid, y_preds),
    'Random Forest': get_metrics(y_valid, rf_preds),
    'XGBoost': get_metrics(y_valid, xgb_preds)
}).T

print(results)

                          MAE        RMSE        R2
Linear Regression  119.899984  162.062298  0.187760
Random Forest       98.256277  137.750162  0.413181
XGBoost             96.437967  136.691303  0.422168


#### Ta thấy XGBoost Model tốt hơn so với 2 Model còn lại:
* XGBoost có tỉ lệ 0.422 nghĩa là Model đã hiểu được khoảng 42.2% lý do tại sao giá đồng hồ lại tăng hay giảm trong khi Linear Regression chỉ có 18,7%
* Con số 42.2% còn quá thấp nên ta quyết định sử dụng XGBoost Model và tìm cách tằn R2 lên

#### Giải thích các thuật ngữ:
* MAE: Sai số tuyệt đối trung bình
* RMSE: Căn bậc hai sai số bình phương trung bình
* R2: Hệ số xác định

In [18]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'n_estimators': [100, 200],           
    'max_depth': [3, 6],                  
    'learning_rate': [0.05, 0.1],         
    'subsample': [0.8, 1.0]  
}             
grid_search = GridSearchCV(
    estimator=xgb.XGBRegressor(random_state=42),
    param_grid=param_grid,
    cv=5,                
    scoring='r2',        
    n_jobs=-1,            
    verbose=1            
)
grid_search.fit(X_train, y_train)

print(f"Cấu hình tốt nhất: {grid_search.best_params_}")
print(f"R2 cao nhất đạt được: {grid_search.best_score_:.4f}")

best_xgb = grid_search.best_estimator_
final_preds_log = best_xgb.predict(X_valid)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
Cấu hình tốt nhất: {'learning_rate': 0.05, 'max_depth': 6, 'n_estimators': 100, 'subsample': 0.8}
R2 cao nhất đạt được: 0.4462


## 4. Tổng kết và Đánh giá (Final Evaluation)

### 4.1. Bảng so sánh kết quả các mô hình
Sau quá trình thực nghiệm, dưới đây là kết quả so sánh giữa các mô hình đã huấn luyện:

| Mô hình | MAE ($) | R² Score | Ghi chú |
| :--- | :---: | :---: | :--- |
| **Linear Regression** | 119.90 | 0.1878 | Độ chính xác thấp, quan hệ phi tuyến tính mạnh |
| **Random Forest** | 98.25 | 0.4131 | Hiệu năng tốt, xử lý được dữ liệu phức tạp |
| **XGBoost (Tuned)** | **96.44** | **0.4462** | **Mô hình tốt nhất (Champion Model)** |

### 4.2. Phân tích kết quả
1. **Sức mạnh của XGBoost:** Mô hình XGBoost sau khi được tối ưu hóa bằng `GridSearchCV` và `5-fold Cross-Validation` đã đạt được hiệu năng ổn định nhất.
2. **Hiện tượng "Trần nhà dữ liệu":** Dù đã áp dụng các kỹ thuật nâng cao (Polynomial Features, Hyperparameter Tuning), chỉ số $R^2$ vẫn đạt ngưỡng bão hòa ở mức **~0.45**. Điều này phản ánh đặc điểm của dữ liệu Marketplace:
   - **Độ nhiễu cao:** Giá bán phụ thuộc nhiều vào cảm quan người bán (tình trạng trầy xước, pin, phụ kiện) - những yếu tố chưa có trong bộ dữ liệu gốc.
   - **Log Transformation:** Việc áp dụng Log giúp giải quyết vấn đề dữ liệu lệch (Skewed), giúp mô hình hội tụ tốt hơn.
